akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -649,3 +652,64 @@ private[remote] class Deserializer(
       setHandlers(in, out, this)
     }
 }
+
+/**
+ * INTERNAL API: The HandshakeReq message must be passed in each inbound lane to
+ * ensure that it arrives before any application message. Otherwise there is a risk
+ * that an application message arrives in the InboundHandshake stage before the
+ * handshake is completed and then it would be dropped.
+ */
+private[remote] class DuplicateHandshakeReq(
+  numberOfLanes:  Int,
+  inboundContext: InboundContext,
+  system:         ExtendedActorSystem,
+  bufferPool:     EnvelopeBufferPool) extends GraphStage[FlowShape[InboundEnvelope, InboundEnvelope]] {
+
+  val in: Inlet[InboundEnvelope] = Inlet("Artery.DuplicateHandshakeReq.in")
+  val out: Outlet[InboundEnvelope] = Outlet("Artery.DuplicateHandshakeReq.out")
+  val shape: FlowShape[InboundEnvelope, InboundEnvelope] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private val (serializerId, manifest) = {
+        val serialization = SerializationExtension(system)
+        val ser = serialization.serializerFor(classOf[HandshakeReq])
+        val m = ser match {
+          case s: SerializerWithStringManifest ⇒
+            s.manifest(HandshakeReq(inboundContext.localAddress, inboundContext.localAddress.address))
+          case _ ⇒ ""
+        }
+        (ser.identifier, m)
+      }
+      var currentIterator: Iterator[InboundEnvelope] = Iterator.empty
+
+      override def onPush(): Unit = {
+        val envelope = grab(in)
+        if (envelope.association.isEmpty && envelope.serializer == serializerId && envelope.classManifest == manifest) {
+          // only need to duplicate HandshakeReq before handshake is completed
+          try {
+            currentIterator = Vector.tabulate(numberOfLanes)(i ⇒ envelope.copyForLane(i)).iterator
+            push(out, currentIterator.next())
+          } finally {
+            val buf = envelope.envelopeBuffer
+            if (buf != null) {
+              envelope.releaseEnvelopeBuffer()
+              bufferPool.release(buf)
+            }
+          }
+        } else
+          push(out, envelope)
+      }
+
+      override def onPull(): Unit = {
+        if (currentIterator.isEmpty)
+          pull(in)
+        else {
+          push(out, currentIterator.next())
+          if (currentIterator.isEmpty) currentIterator = Iterator.empty // GC friendly
```

Comment:
Same here, the iterator never becomes non-empty again after becoming empty, right? So maybe both change handlers...